File: 2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level.py
from collections import deque

class Solution:
    def minSteps(self, level):
        n = len(level)
        eleInd = [(val, i) for i, val in enumerate(level)]
        eleInd.sort()

        vis = [0] * n
        ans = 0

        for i in range(n):
            if vis[i] or eleInd[i][1] == i:
                continue

            cycle = 0
            j = i

            while not vis[j]:
                vis[j] = 1
                j = eleInd[j][1]
                cycle += 1

            ans += cycle - 1

        return ans

    def minimumOperations(self, root):
        if not root:
            return 0

        ans = 0
        q = deque([root])

        while q:
            level = []
            for _ in range(len(q)):
                node = q.popleft()
                level.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            ans += self.minSteps(level)

        return ans

File: 2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2558-minimum-number-of-operations-to-sort-a-binary-tree-by-level.rb
    def min_steps(level)
        n = level.size
        ele_ind = level.each_with_index.map { |val, i| [val, i] }
        ele_ind.sort_by! { |x| x[0] }

        vis = Array.new(n, false)
        ans = 0

        (0...n).each do |i|
            next if vis[i] || ele_ind[i][1] == i

            cycle = 0
            j = i

            while !vis[j]
                vis[j] = true
                j = ele_ind[j][1]
                cycle += 1
            end

            ans += cycle - 1
        end

        ans
    end

    def minimum_operations(root)
        return 0 if root.nil?

        ans = 0
        q = [root]

        until q.empty?
            level = []
            next_level = []

            q.each do |node|
                level << node.val
                next_level << node.left if node.left
                next_level << node.right if node.right
            end

            q = next_level
            ans += min_steps(level)
        end

        ans
    end
